Bjerksund-Stensland 2002 option pricing model

alembcke opened the pull request:
This pull requests adds the Bjerksund-Stensland 2002 option pricing model, which closes #5.

Tests are included which mirror the tables in the [reference paper](https://core.ac.uk/download/pdf/30824897.pdf).  Additional testing in both eager and deferred (graph) mode were conducted with visual inspection of the output of the code.  All tests were successful.

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -350,7 +350,7 @@ def bjerksund_stensland(*,
-                        modified_boundary: bool = False,
+                        modified_boundary: bool = True,
                         dtype: tf.DType = None,
                         name=None) -> types.RealTensor:
   """Computes prices of a batch of American options using Bjerksund-Stensland.
```
please update Raises section of the function. Also, add 

name: str = None

alembcke replied:
Updated.  I also updated the example code, as it was no longer relevant.

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -601,5 +681,63 @@ def _npdf(x):
+
+  dtype = rho.dtype
+
+  # Gauss Legendre Points and Weights
+  w = tf.constant([0.01761400713915212, 0.04060142980038694, 0.06267204833410906,
```
Please add a comment that unlike in the reference, for efficiency reasons on the GPU, we use 10 integration points regardless of `rho` values

alembcke replied:
Updated.

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option_test.py:
```diff
@@ -338,6 +338,139 @@ def test_bs1993_prices_long_term2(self, discount_rates, dividend_rates,
                           msg=msg)
+  
+  @parameterized.parameters(
+      (0.08, 0.12, 0.2, 0.25,
```
could you please add a test where all inputs are ternsors and where some options are puts and some are calls?

alembcke replied:
Added.  Let me know if this isn't quite what you were looking for.

cyrilchim replied:
Sorry, I did not explain myself well. I meant, at the moment rates and expiries are scalars. Could you make them tensors of the same shape as spots. E.g., for discount rates `discount_rates` use `2 * [0.08, 0.07, 0.06, 0.05, 0.04]`. Just one test will do. This way we know that batching works.

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -601,5 +681,63 @@ def _npdf(x):
+  hs = (h * h + k * k) / 2
+  asr = tf.math.asin(rho)
+
+  def batched_calc(hk, hs, asr, x, w):
+      for _ in range(hk.shape.rank):
```
could we maybe have any better name for this function? Does it correspond to anything from the paper/code?

alembcke replied:
In the paper it is called "The Transformed BVN Problem" in section 2.2 equation 3, while the original is called "The Standard BVN Problem" in section 2.1 equation 1.  The transformation is that the double-integral is transformed into a single integral with the substitution `r = sin(theta)`.  I renamed the function `transformed_bvn`, is that suitable?

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -601,5 +681,63 @@ def _npdf(x):
+        # Shape x.shape + [1, .., 1]
+        x = tf.expand_dims(x, axis=-1)
+        w = tf.expand_dims(w, axis=-1)
+      # Shape x.shape + asr.shape
+      sn1 = tf.math.sin(asr * (-1 * x + 1) / 2)
```
instead of `asr.shape`, let us have `bvn.shape`

alembcke replied:
Do you mean update the comments to say that?  I assumed so and updated accordingly, but let me know if I got that wrong.

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -601,5 +681,63 @@ def _npdf(x):
+
+  def batched_calc(hk, hs, asr, x, w):
+      for _ in range(hk.shape.rank):
+        # Shape x.shape + [1, .., 1]
+        x = tf.expand_dims(x, axis=-1)
```
# Shape x.shape + bvn.shape.rank * [1]

alembcke replied:
Updated.

cyrilchim reviewed (commented):
Just a few minor comments. Otherwise, LGTM

alembcke commented:
Made the requested changes, see unresolved comments above for further clarification.

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -350,9 +350,9 @@ def bjerksund_stensland(*,
                         dividend_rates: types.RealTensor = None,
                         discount_factors: types.RealTensor = None,
```
could you please squash the commits?

alembcke replied:
Sure.

cyrilchim reviewed (commented):
Thanks! Just one final comment. Please squash the commits as well.

alembcke commented:
Updated the test so that `discount_rates`, `dividend_rates` and `expiries` are not scalars, to test batching.  Tests ran successfully.  Also, I squashed the commits.

cyrilchim commented:
Thank you Alex! This is now merged

aminehadbi commented:
Note to self to research this.
